### PR TITLE
APP-2198: Config page: add component creation success toast

### PR DESCRIPTION
**Pull request by Oabraham1**
[Ticket](https://viam.atlassian.net/browse/APP-2198)

The goal of this PR is two-fold: Implement a `Toast` and display that `Toast` on successful component creation.

## Screenshot
Screenshot 2023-12-15 at 14 48 05

### CODE REVIEW

File: packages/core/.vscode/settings.json
Comment (DTCurrie) [inline packages/core/.vscode/settings.json:2]: Did we intend to commit this?

**Comment by DTCurrie** [comment]
Can you include screenshots of what this component looks like?

File: packages/core/src/lib/toast/context.ts
Comment (DTCurrie) [inline packages/core/src/lib/toast/context.ts:57]: Not an issue, but why a `Map` over a `Record`?

File: packages/core/src/lib/toast/context.ts
Comment (DTCurrie) [inline packages/core/src/lib/toast/context.ts:93]: Why are we doing this `get` first?

**Comment by ethanlook** [comment]
> Can you include screenshots of what this component looks like?

+1 for [APP-2847](https://viam.atlassian.net/browse/APP-2847)

File: packages/core/src/lib/toast/context.ts
Comment (DTCurrie) [inline packages/core/src/lib/toast/context.ts:26]: I feel like `upload` is a pretty specific use case for this, should it be something like `info` or `progress`? As a user, I have no idea what an `upload` toast is supposed to look like or signify.

File: packages/core/src/lib/toast/toast.svelte
Comment (DTCurrie) [inline packages/core/src/lib/toast/toast.svelte:35]: ✂️

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (DTCurrie) [inline packages/core/src/lib/toast/toast-banner.svelte:93]: This kind of feels like it should be a `slot`, maybe wrapped in a layout element with the `"absolute right-1"` classes. Especially if all we are doing is having it be an event dispatcher, might as well just let users declare exactly what they need. Will make it more flexible for cases where we may just want to use an icon or something in the future as well.

File: packages/core/src/lib/toast/context.ts
Comment (ethanlook) [reply packages/core/src/lib/toast/context.ts:26]: I pulled down the PR and ran the storybook and it looks like this is intended to be the "process" variant we have in the [figma](https://www.figma.com/file/ERHSYDGTXiEQOVGY8ORg6I/PRIME?type=design&node-id=2650-4838&mode=design&t=hht9SlFQ2x7ltRIa-4).

If this is supposed to be the "process" variant, a few notes:
- Progress bar should go left-to-right
- Process variant should end looking like "success"
- Progress bar should have controlled progress such that it reflects the progress of the long-running action (such as uploading a file)

If this is supposed to be the "info" variant:
- Icon should be the info icon
- There should not be a progress bar

Which [variants](https://www.figma.com/file/ERHSYDGTXiEQOVGY8ORg6I/PRIME?type=design&node-id=2650-4838&mode=design&t=hht9SlFQ2x7ltRIa-4) do we need for R2D2?

File: packages/core/.vscode/settings.json
Comment (DTCurrie) [reply packages/core/.vscode/settings.json:2]: But `exitable` is not a word, and I can't see where we are using it.

File: packages/core/src/lib/toast/context.ts
Comment (DTCurrie) [reply packages/core/src/lib/toast/context.ts:26]: `process` definitely makes sense as a variant name. I am wondering if we even need the process variant right now? The ticket this is attached to is specifically for the success toast, so seems like we could hold off on the `process` variant right now?

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [reply packages/core/src/lib/toast/context.ts:57]: Since we're dynamically adding and removing these handlers, I think a `Map` is better than a plain-object record. We have lint rules that will yell at you for using `delete obj[key]`

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [reply packages/core/src/lib/toast/context.ts:93]: To get and call the unsubscribe function, if needed, before deleting the unsubscribe function

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [inline packages/core/src/lib/toast/context.ts:24]: Same question as below with `upload/process`. Our immediate needs with toasts are all success toasts. Let's not add anything in this PR that we aren't going to use immediately

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [inline packages/core/src/lib/toast/context.ts:41]: Same comment as elsewhere: all toasts that we care about right now are closable

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-banner.svelte:16]: Avoid using empty strings as falsey values; use `undefined` to declare that something might not be specified. It's a lot easier to statically analyze

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-banner.svelte:34]: I know this was copied over from the the notification code, but can we use function props rather than events? Svelte itself is moving away from events, and in my experience, events are annoying to test for and statically analyze

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-banner.svelte:19]: Without a `process` variant, do we need this prop?

File: packages/storybook/src/stories/use-toast.mdx
Comment (mcous) [inline packages/storybook/src/stories/use-toast.mdx:59]: I know this comes from the `notify` usage, but I think it's a great time to reimagine this. For me, the level seems better as an options argument rather than having separate methods for each level

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-item.svelte:16]: We _do_ need actions to work for R2D2, so this looks like an issue

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [reply packages/core/src/lib/toast/context.ts:24]: Agree, I think we need an options object with the following fields:

- `message: string`
- `action: { text: string, handler: () => unknown }`
- `variant: ToastVariant`

As mentioned in other comments, `closable` should not be an option; the closable state of a toast is determined by its variant

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-item.svelte:1]: Is this component needed? Can it be subsumed into `<ToastContainer>`? It doesn't have any specific tests and it doesn't do anything meaningful other than apply CSS

File: packages/core/.vscode/settings.json
Comment (mcous) [inline packages/core/.vscode/settings.json:1]: Setting aside the conversation below, this repository is designed to be opened in your editor from the root, so any `.vscode` changes should be committed at `<root>/.vscode` rather than `<root>/packages/[package]/.vscode`. This directory should be (re)moved

**Review by mcous — CHANGES_REQUESTED** [review]

File: packages/core/src/lib/toast/context.ts
Comment (DTCurrie) [inline packages/core/src/lib/toast/context.ts:141]: Open question for reviewers, do we prefer full syntax here, or just one liners like so?

```suggestion
/** Get access to the toast notifier in a component. */
export const useToast = (): Toast => useToastContext().toast;

/** Get access to the internal toast state. */
export const useToastState = (): ToastState => useToastContext().state;
```

File: packages/storybook/src/stories/use-toast.mdx
Comment (DTCurrie) [reply packages/storybook/src/stories/use-toast.mdx:59]: So update the API to something like this?

```
  toast('Plain slice please!');
  toast('Uploading file!', ToastVariant.Upload);
  toast('Fuhgeddaboudit!', ToastVariant.Success);
```

File: packages/core/src/lib/toast/context.ts
Comment (ethanlook) [reply packages/core/src/lib/toast/context.ts:141]: I think I prefer

```
const { state, toast } = useToast();
```

Comment (mcous) [reply packages/core/src/lib/toast/context.ts:141]: One wrinkle: toast state is internal

File: packages/core/src/lib/toast/toast.svelte
Comment (mrloureed) [inline packages/core/src/lib/toast/toast.svelte:91]: I don't understand `-mr-py -mb-py`

File: packages/core/src/lib/toast/context.ts
Comment (Oabraham1) [reply packages/core/src/lib/toast/context.ts:26]: Will focus this PR on the success variant.

File: packages/core/src/lib/toast/toast.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast.svelte:35]: Removing this file from the PR as it was added in error.

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mrloureed) [inline packages/core/src/lib/toast/toast-item.svelte:14]: You probably don't need the `pointer-events-auto`

Nice use of `sm:` utility

File: packages/core/src/lib/toast/toast-container.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-container.svelte:31]: I actually was going to remove both since we do not want the progress of the toast to be paused when a user hovers over it right?

File: packages/core/src/lib/toast/toast-item.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-item.svelte:6]: Yes, I think I was using this as a way whether to determine whether or not a close icon button is displayed, but I'm taking it out

File: packages/storybook/src/stories/use-toast.mdx
Comment (mcous) [reply packages/storybook/src/stories/use-toast.mdx:59]: Yeah, though I'd use an options object, as Ethan suggested in his review. So something like:

```ts
toast({ message: 'Hello world', variant: ToastVariant.SUCCESS })
```

From there, if we wanted, we could layer stuff in like "default to `INFO`" and "allow a plain string to toast an info message"

File: packages/core/src/lib/toast/toast.svelte
Comment (mcous) [reply packages/core/src/lib/toast/toast.svelte:91]: From what I can tell, they're not tailwind classes. The fact that they're shunted to the front is a warning sign; that's where prettier puts non-tailwind classes

(Fun fact: our linter isn't capable of warning us about non-tailwind classes because that part of eslint-plugin-tailwindcss [isn't fully compatible with eslint-plugin-svelte](https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/214). Yay.)

File: packages/core/src/lib/toast/toast-item.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-item.svelte:16]: I will modify this. Question: does the success toast we're using use actions in any way? From my understanding, only the close button is displayed on a success toast. I'm asking because I might make the action field optional

Comment (mcous) [reply packages/core/src/lib/toast/toast-item.svelte:16]: Yes, various success toasts for destructive actions include an "Undo" action

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-banner.svelte:19]: Removing it

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:17]: - Use data-testid for testing slots, as described above
- Try to always separate your element queries from you expectations for readability
- Try to always use names when you're querying buttons, especially when they're icon buttons, which need extra a11y care
- Almost never use `container` and almost never use `querySelector`
    - Needing to resort to these is usually a sign that your HTML isn't properly accessible and/or semantic
    - In this case, I would say that the `Icon` needs to be "visible" to the a11y tree as an image with a label of "success," because it conveys meaning for the toast message that isn't available elsewhere

```suggestion
    const message = screen.getByTestId('message')
    const action = screen.getByTestId('action')
    const icon = screen.getByRole('img', { name: /success/iu })
    const dismissButton = screen.getByRole('button', { name: /dismiss/iu })
    
    expect(message).toBeVisible();
    expect(action).toBeVisible();
    expect(icon).toHaveClass('text-success-dark');
    expect(dismissButton).toBeVisible();
```

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:21]: [Nit] The "mock" in the name here is a bit redundant; it's defined with `vi.fn()`, we know what it is. `dismiss` is a little easier to read/use

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:26]: - Separate your queries from user actions
- `getByRole` is almost always preferred if you can use it

```suggestion
    const dismissButton = screen.getByRole('button', { name: /dismiss/iu });
    
    await fireEvent.click(dismissButton);
```

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:37]: This test smells due to the `query-selector`, and what it's testing is a bit of a visual implementation detail. Unless there's a good semantic reason to test a classname (e.g. something representing success has a class to apply our "text-success-dark", something needs to be hidden) I'd recommend avoiding these sorts of tests, _especially_ for internal components like `<ToastBanner />`

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:1]: Some of the test coverage in this file looks redundant with `toast.spec.ts`. Since `toast.spec.ts` tests the actual public interface of the `toast` lib, that's where I'd recommend putting all your tests

These sort of internal unit test _can_ be useful, but constraining yourself to public interfaces creates less coupling between your tests and your internal implementation details, which is almost always preferable if you can manage it

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mcous) [reply packages/core/src/lib/toast/toast-item.svelte:1]: Bumping this comment. It looks like splitting apart `ToastBanner` and `ToastItem` is creating a more complicated internal structure for this functionality than is needed

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-banner.svelte:26]: This logic looks like it could live in `variants.ts` and be a little easier to read

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-item.svelte:22]: This `if` looks too low, this usage of `slot="action"` is going to trigger `if $$slots.action` in `<ToastBanner>`. I think this mostly relates to me comment above about there being one two many components in play in this internal structure

File: packages/core/src/lib/toast/toast-item.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-item.svelte:15]: Prefer function prop to event emitter

File: packages/core/src/lib/toast/context.ts
Comment (mcous) [inline packages/core/src/lib/toast/context.ts:127]:
```suggestion
export const useToast = (): ((params: ToastParams) => void) => {
  return useToastContext().toast;
};
```

File: packages/core/src/lib/toast/__tests__/toast-banner.spec.ts
Comment (Oabraham1) [reply packages/core/src/lib/toast/__tests__/toast-banner.spec.ts:38]: I'll remove the test case :)

File: packages/core/src/lib/toast/toast-item.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-item.svelte:22]: I can move everything in this file to <ToastBanner>

**Comment by Oabraham1** [comment]
@mcous please let me know if I missed any of your comments, but I think I addressed everything.

File: packages/core/src/lib/toast/__tests__/toast.spec.ts
Comment (mcous) [inline packages/core/src/lib/toast/__tests__/toast.spec.ts:62]: Prefer `@testing-library/user-event` to `fireEvent`

```ts
import userEvent from '@testing-library/user-event'

// sometime before render
const user = userEvent.setup()

// click something
await user.click(element)
```

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-banner.svelte:36]: [Nit] Is `pointer-events-auto` doing anything here?

File: packages/core/src/lib/toast/toast-container.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-container.svelte:41]: [Nit] I'm not sure `inline-block` does anything when the parent is `flex`

File: packages/core/src/lib/toast/variants.ts
Comment (mcous) [inline packages/core/src/lib/toast/variants.ts:3]: [Nit]: let's remove these while we're not using them

File: packages/core/src/lib/toast/toast-container.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-container.svelte:31]: We're missing an `overflow-hidden` on this element; in the storybook, a scrollbar appears while toasts are flying in if you're on desktop with a mouse (on macOS, it's likely worse on Windows)

[Nit] I think `z-[9999]` is equivalent to our `z-max`

File: packages/core/src/lib/toast/toast-container.svelte
Comment (mcous) [inline packages/core/src/lib/toast/toast-container.svelte:37]: Is it just me, or are the elements not actually fading out? On my machine they don't fade until I remove the `inline-block` class below

**Review by mcous — COMMENTED** [review]
Almost there! Got a couple small issues with the CSS / animations, but we'll be good to merge once they are resolved

File: packages/core/src/lib/toast/toast-banner.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-banner.svelte:36]: Removed it since the one from `toast-container.svelte` is used to pause the toast progress on hover

File: packages/core/src/lib/toast/toast-container.svelte
Comment (Oabraham1) [reply packages/core/src/lib/toast/toast-container.svelte:41]: I removed it.

**Review by mcous — APPROVED** [review]

**Pull request merged**
